compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,165 @@
+        "SimplifyStaticSwitch"
+    }
+
+    fn is_enabled(&self, sess: &Session) -> bool {
+        sess.mir_opt_level() >= 2
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!(body = ?body.source.def_id());
+        if simplify_static_switch(tcx, body) {
+            remove_dead_blocks(tcx, body);
+        }
+    }
+}
+
+#[instrument(level = "debug", skip(tcx, body), ret)]
+fn simplify_static_switch<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) -> bool {
+    let predecessors = body.basic_blocks.predecessors();
+    let borrowed_locals = borrowed_locals(body);
+    let mut new_targets = FxHashMap::default();
+    for (block, data) in body.basic_blocks.iter_enumerated() {
+        if data.is_cleanup {
+            continue;
+        }
+
+        let predecessors = &predecessors[block];
+        if predecessors.is_empty() {
+            continue;
+        }
+
+        let TerminatorKind::SwitchInt {
+            discr: Operand::Move(switched) | Operand::Copy(switched),
+            ref targets
+        } = data.terminator().kind else {
+            continue
+        };
+
+        let Some(switched) = switched.as_local() else {
+            continue
+        };
+
+        if borrowed_locals.contains(switched) {
+            continue;
+        }
+
+        let Some(discr) = data.statements.iter().rev().find_map(|statement| {
+            match statement.kind {
+                StatementKind::Assign(box (place, Rvalue::Discriminant(discr)))
+                    if place.as_local() == Some(switched)
+                        && let Some(local) = discr.as_local()
+                        && !borrowed_locals.contains(local) =>
+                {
+                    Some(discr)
+                },
+                _ => None,
+            }
+        }) else {
+            continue
+        };
+
+        for &pred in predecessors {
+            let data = &body.basic_blocks[pred];
+            let TerminatorKind::Goto { .. } = data.terminator().kind else {
+                continue
+            };
+
+            for statement in data.statements.iter().rev() {
+                match statement.kind {
+                    StatementKind::Assign(box (
+                        place,
+                        Rvalue::Aggregate(box AggregateKind::Adt(def_id, variant, substs, ..), ..),
+                    )) if place == discr => {
+                        debug!("{pred:?}: {place:?} = {def_id:?}::{variant:?}; goto -> {block:?};");
+                        let discr_ty = tcx.type_of(def_id).subst(tcx, substs);
+                        if let Some(discr) = discr_ty.discriminant_for_variant(tcx, variant) {
+                            let target = targets.target_for_value(discr.val);
+                            new_targets
+                                .entry((place, block))
+                                .or_insert_with(Vec::new)
+                                .push((pred, target));
```

Comment:
There is a subtle invariant here: that the `(pred, target)` edge only appears once in the cfg. This is fulfilled in your implementation as `pred` ends with a `goto`, but this deserves to be commented.

Reply:
i added a comment mentioning this.